File: server/arduinoSerial.ts
// Task #2: Serial Communication Handler for Arduino
import { SerialPort } from 'serialport';
import { ReadlineParser } from '@serialport/parser-readline';
import SensorData from './models/SensorData.js';
import Rule from './models/Rule.js';

export interface SensorReading {
  gas: number;
  light: number;
  soil: number;
  water: number;
  infrar: number;
  btn1: number;
  btn2: number;
  timestamp: Date;
}

export interface ActuatorStates {
  white_light: boolean;
  yellow_light: boolean;
  relay: boolean;
  door_angle: number;
  window_angle: number;
  fan: boolean;
  fan_speed: number;
  buzzer: boolean;
}

export class ArduinoSerial {
  private port: SerialPort | null = null;
  private parser: ReadlineParser | null = null;
  private currentData: SensorReading | null = null;
  private actuatorStates: ActuatorStates = {
    white_light: false,
    yellow_light: false,
    relay: false,
    door_angle: 0,
    window_angle: 0,
    fan: false,
    fan_speed: 0,
    buzzer: false
  };
  private dataBuffer: SensorReading[] = [];
  private saveInterval: NodeJS.Timeout | null = null;
  private isReady = true;
  private commandQueue: string[] = [];
  constructor() {
    // Start saving data every 2 seconds
    this.saveInterval = setInterval(() => {
      this.saveBufferedData();
    }, 2000);
  }

  static async listPorts(): Promise<any[]> {
    const { SerialPort } = await import('serialport');
    return SerialPort.list();
  }

  async connect(portPath: string, baudRate: number = 9600): Promise<void> {
    try {
      if (this.port && this.port.isOpen) {
        console.warn('Port already open');
        return;
      }

      // Open serial port
      this.port = new SerialPort({ path: portPath, baudRate });

      // Wait 2 seconds for Arduino to reset after port open
      await new Promise(res => setTimeout(res, 2000));

      // Set up parser for newline-delimited data
      this.parser = this.port.pipe(new ReadlineParser({ delimiter: '\n' }));

      // Set up listeners
      this.setupListeners();

      console.log(`✓ Connected to Arduino on ${portPath}`);
    } catch (error) {
      console.error('Failed to connect to Arduino:', error);
      throw error;
    }
  }

  async disconnect(): Promise<void> {
    if (this.saveInterval) {
      clearInterval(this.saveInterval);
      this.saveInterval = null;
    }
    
    if (this.port && this.port.isOpen) {
      await new Promise<void>((resolve) => {
        this.port!.close(() => {
          console.log('✓ Disconnected from Arduino');
          resolve();
        });
      });
    }
    
    this.port = null;
    this.parser = null;
  }

  private setupListeners(): void {
    if (!this.parser || !this.port) return;

    // Parser listens for each line from Arduino
    this.parser.on('data', (line: string) => {
      const trimmed = line.trim();
      console.log('←', trimmed);

      // ACK handling for command queue
      if (trimmed === 'ACK') {
        this.isReady = true;

        // Send next command if queued
        if (this.commandQueue.length > 0) {
          const nextCommand = this.commandQueue.shift()!;
          setTimeout(() => this.sendCommand(nextCommand), 100); // slight delay
        }
        return;
      }

      // Process sensor or actuator data
      this.handleIncomingData(trimmed);
    });

    // Handle parser errors
    this.parser.on('error', (error) => {
      console.error('Serial parser error:', error);
    });

    // Handle port-level errors
    this.port.on('error', (error) => {
      console.error('Serial port error:', error);
    });
  }


  private handleIncomingData(line: string): void {
    try {
      // Parse Arduino sensor data format: "GAS:123,LIGHT:456,SOIL:789,WATER:101,INFRAR:1,BTN1:0,BTN2:1"
      if (line.includes('GAS:')) {
        const sensorData = this.parseSensorData(line);
        if (sensorData) {
          this.currentData = sensorData;
          this.dataBuffer.push(sensorData);
          
          // Keep buffer size manageable
          if (this.dataBuffer.length > 100) {
            this.dataBuffer = this.dataBuffer.slice(-50);
          }
          
          // Evaluate rules for edge analytics
          this.evaluateRules(sensorData);
        }
      }
      // Handle actuator acknowledgments
      else if (line.includes('light ON') || line.includes('light OFF') || 
               line.includes('opened') || line.includes('closed') ||
               line.includes('Fan ON') || line.includes('Fan OFF') ||
               line.includes('Relay ON') || line.includes('Relay OFF') ||
               line.includes('Buzzer ON') || line.includes('Buzzer OFF')) {
        this.updateActuatorStateFromResponse(line);
      }
    } catch (error) {
      console.error('Error handling incoming data:', error);
    }
  }

  private parseSensorData(line: string): SensorReading | null {
    try {
      const parts = line.split(',');
      const data: any = {};
      
      parts.forEach(part => {
        const [key, value] = part.split(':');
        if (key && value !== undefined) {
          data[key.toLowerCase()] = parseInt(value, 10);
        }
      });

      if (data.gas !== undefined && data.light !== undefined && 
          data.soil !== undefined && data.water !== undefined) {
        return {
          gas: data.gas,
          light: data.light,
          soil: data.soil,
          water: data.water,
          infrar: data.infrar || 0,
          btn1: data.btn1 || 0,
          btn2: data.btn2 || 0,
          timestamp: new Date()
        };
      }
      
      return null;
    } catch (error) {
      console.error('Error parsing sensor data:', error);
      return null;
    }
  }

  private updateActuatorStateFromResponse(response: string): void {
    if (response.includes('White light ON')) {
      this.actuatorStates.white_light = true;
    } else if (response.includes('White light OFF')) {
      this.actuatorStates.white_light = false;
    } else if (response.includes('Yellow light ON')) {
      this.actuatorStates.yellow_light = true;
    } else if (response.includes('Yellow light OFF')) {
      this.actuatorStates.yellow_light = false;
    } else if (response.includes('Relay ON')) {
      this.actuatorStates.relay = true;
    } else if (response.includes('Relay OFF')) {
      this.actuatorStates.relay = false;
    } else if (response.includes('Door opened')) {
      this.actuatorStates.door_angle = 180;
    } else if (response.includes('Door closed')) {
      this.actuatorStates.door_angle = 0;
    } else if (response.includes('Window opened')) {
      this.actuatorStates.window_angle = 180;
    } else if (response.includes('Window closed')) {
      this.actuatorStates.window_angle = 0;
    } else if (response.includes('Fan ON')) {
      this.actuatorStates.fan = true;
    } else if (response.includes('Fan OFF')) {
      this.actuatorStates.fan = false;
    } else if (response.includes('Buzzer ON')) {
      this.actuatorStates.buzzer = true;
    } else if (response.includes('Buzzer OFF')) {
      this.actuatorStates.buzzer = false;
    } else if (response.includes('Door angle:')) {
      const angle = parseInt(response.split(':')[1]?.trim() || '0', 10);
      this.actuatorStates.door_angle = angle;
    } else if (response.includes('Window angle:')) {
      const angle = parseInt(response.split(':')[1]?.trim() || '0', 10);
      this.actuatorStates.window_angle = angle;
    } else if (response.includes('Fan speed:')) {
      const speed = parseInt(response.split(':')[1]?.trim() || '0', 10);
      this.actuatorStates.fan_speed = speed;
    }
  }

  private async evaluateRules(sensorData: SensorReading): Promise<void> {
    try {
      const rules = await Rule.find({ enabled: true });
      
      for (const rule of rules) {
        const sensorValue = sensorData[rule.sensor as keyof SensorReading] as number;
        let triggered = false;
        
        switch (rule.operator as any) {
          case '>':
            triggered = sensorValue > rule.threshold;
            break;
          case '<':
            triggered = sensorValue < rule.threshold;
            break;
          case '>=':
            triggered = sensorValue >= rule.threshold;
            break;
          case '<=':
            triggered = sensorValue <= rule.threshold;
            break;
          case '==':
            triggered = sensorValue === rule.threshold;
            break;
        }
        
        if (triggered) {
          console.log(`🔥 Rule triggered: ${rule.name} (${rule.sensor}: ${sensorValue} ${rule.operator} ${rule.threshold})`);
          await this.executeAction(rule.action, sensorData);
        }
      }
    } catch (error) {
      console.error('Error evaluating rules:', error);
    }
  }

  private async executeAction(action: string, sensorData: SensorReading): Promise<void> {
    try {
      switch (action) {
        case 'white_light_on':
          if (!this.actuatorStates.white_light) {
            this.sendCommand('white_light_on');
          }
          break;
        case 'white_light_off':
          if (this.actuatorStates.white_light) {
            this.sendCommand('white_light_off');
          }
          break;
        case 'yellow_light_on':
          if (!this.actuatorStates.yellow_light) {
            this.sendCommand('yellow_light_on');
          }
          break;
        case 'yellow_light_off':
          if (this.actuatorStates.yellow_light) {
            this.sendCommand('yellow_light_off');
          }
          break;
        case 'relay_on':
          if (!this.actuatorStates.relay) {
            this.sendCommand('relay_on');
          }
          break;
        case 'relay_off':
          if (this.actuatorStates.relay) {
            this.sendCommand('relay_off');
          }
          break;
        case 'door_open':
          if (this.actuatorStates.door_angle !== 180) {
            this.sendCommand('door_open');
          }
          break;
        case 'door_close':
          if (this.actuatorStates.door_angle !== 0) {
            this.sendCommand('door_close');
          }
          break;
        case 'window_open':
          if (this.actuatorStates.window_angle !== 180) {
            this.sendCommand('window_open');
          }
          break;
        case 'window_close':
          if (this.actuatorStates.window_angle !== 0) {
            this.sendCommand('window_close');
          }
          break;
        case 'fan_on':
          if (!this.actuatorStates.fan) {
            this.sendCommand('fan_on');
          }
          break;
        case 'fan_off':
          if (this.actuatorStates.fan) {
            this.sendCommand('fan_off');
          }
          break;
        case 'buzzer_on':
          this.sendCommand('buzzer_on');
          // Auto turn off buzzer after 2 seconds
          setTimeout(() => {
            this.sendCommand('buzzer_off');
          }, 2000);
          break;
        case 'buzzer_off':
          if (this.actuatorStates.buzzer) {
            this.sendCommand('buzzer_off');
          }
          break;
        default:
          console.warn(`Unknown action: ${action}`);
      }
      
      // Save alert to database
      if (this.currentData) {
        const alertData = {
          ...this.currentData,
          alerts: [`Action executed: ${action}`]
        };
        
        const sensorDoc = new SensorData(alertData);
        await sensorDoc.save();
      }
    } catch (error) {
      console.error('Error executing action:', error);
    }
  }

  sendCommand(command: string): void {
    if (!this.port || !this.port.isOpen) {
      console.warn('Cannot send command: Arduino not connected');
      return;
    }

    // Queue command if Arduino is busy
    if (!this.isReady) {
      this.commandQueue.push(command);
      console.log(`⏳ Queued: ${command}`);
      return;
    }

    // Send immediately if ready
    this.isReady = false;
    this.port.write(command + '\n', (err) => {
      if (err) console.error('Error writing to port:', err);
      else console.log(`→ Sent: ${command}`);
    });
  }


  getCurrentData(): SensorReading | null {
    return this.currentData;
  }

  getActuatorStates(): ActuatorStates {
    return { ...this.actuatorStates };
  }

  setActuatorState(actuator: string, value: any): void {
    if (actuator in this.actuatorStates) {
      (this.actuatorStates as any)[actuator] = value;
    }
  }

  private async saveBufferedData(): Promise<void> {
    if (this.dataBuffer.length === 0) return;
    
    try {
      const dataToSave = this.dataBuffer.splice(0);
      
      if (dataToSave.length > 0) {
        await SensorData.insertMany(dataToSave);
        console.log(`💾 Saved ${dataToSave.length} sensor readings to database`);
      }
    } catch (error) {
      console.error('Error saving sensor data:', error);
    }
  }
}

File: server/serial/arduino_serial.go
package serial

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/caphefalumi/smart-home/server/models"
	"github.com/caphefalumi/smart-home/server/services"
	"github.com/tarm/serial"
)

// ArduinoSerial handles serial communication with Arduino
type ArduinoSerial struct {
	port           io.ReadWriteCloser
	isConnected    bool
	currentData    *models.SensorReading
	actuatorStates *models.ActuatorStates
	dataBuffer     []models.SensorData
	sensorService  *services.SensorService
	ruleService    *services.RuleService
	commandQueue   []string
	isReady        bool
	mutex          sync.RWMutex
	stopChan       chan bool
	saveInterval   time.Duration
}

// NewArduinoSerial creates a new Arduino serial handler
func NewArduinoSerial(sensorService *services.SensorService, ruleService *services.RuleService) *ArduinoSerial {
	return &ArduinoSerial{
		actuatorStates: &models.ActuatorStates{},
		dataBuffer:     make([]models.SensorData, 0),
		sensorService:  sensorService,
		ruleService:    ruleService,
		commandQueue:   make([]string, 0),
		isReady:        true,
		stopChan:       make(chan bool),
		saveInterval:   2 * time.Second,
	}
}

// ListPorts returns available serial ports
func (a *ArduinoSerial) ListPorts() ([]string, error) {
	// On Windows, common ports are COM1, COM2, etc.
	// This is a simplified implementation - in production you'd use a proper port discovery
	commonPorts := []string{"COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7", "COM8", "COM9", "COM10"}

	var availablePorts []string
	for _, portName := range commonPorts {
		config := &serial.Config{
			Name:        portName,
			Baud:        9600,
			ReadTimeout: time.Millisecond * 100,
		}

		if port, err := serial.OpenPort(config); err == nil {
			port.Close()
			availablePorts = append(availablePorts, portName)
		}
	}

	return availablePorts, nil
}

// Connect establishes connection to Arduino
func (a *ArduinoSerial) Connect(portName string, baudRate int) error {
	a.mutex.Lock()
	defer a.mutex.Unlock()

	if a.isConnected {
		return fmt.Errorf("already connected to a port")
	}

	config := &serial.Config{
		Name:        portName,
		Baud:        baudRate,
		ReadTimeout: time.Second * 5,
	}

	port, err := serial.OpenPort(config)
	if err != nil {
		return fmt.Errorf("failed to open port %s: %w", portName, err)
	}

	a.port = port
	a.isConnected = true

	// Wait for Arduino to reset
	time.Sleep(2 * time.Second)

	// Start data processing goroutines
	go a.startDataListener()
	go a.startDataSaver()

	log.Printf("✓ Connected to Arduino on %s", portName)
	return nil
}

// Disconnect closes the serial connection
func (a *ArduinoSerial) Disconnect() error {
	a.mutex.Lock()
	defer a.mutex.Unlock()

	if !a.isConnected {
		return nil
	}

	// Stop background processes
	close(a.stopChan)

	// Save any remaining buffered data
	a.saveBufferedData()

	if a.port != nil {
		a.port.Close()
		a.port = nil
	}

	a.isConnected = false
	log.Println("✓ Disconnected from Arduino")
	return nil
}

// IsConnected returns connection status
func (a *ArduinoSerial) IsConnected() bool {
	a.mutex.RLock()
	defer a.mutex.RUnlock()
	return a.isConnected
}

// SendCommand sends a command to Arduino
func (a *ArduinoSerial) SendCommand(command string) error {
	a.mutex.Lock()
	defer a.mutex.Unlock()

	if !a.isConnected || a.port == nil {
		return fmt.Errorf("not connected to Arduino")
	}

	if !a.isReady {
		// Queue the command
		a.commandQueue = append(a.commandQueue, command)
		return nil
	}

	// Send command immediately
	_, err := a.port.Write([]byte(command + "\n"))
	if err != nil {
		return fmt.Errorf("failed to send command: %w", err)
	}

	a.isReady = false
	log.Printf("→ %s", command)
	return nil
}

// GetCurrentData returns the latest sensor reading
func (a *ArduinoSerial) GetCurrentData() *models.SensorReading {
	a.mutex.RLock()
	defer a.mutex.RUnlock()
	return a.currentData
}

// GetActuatorStates returns current actuator states
func (a *ArduinoSerial) GetActuatorStates() *models.ActuatorStates {
	a.mutex.RLock()
	defer a.mutex.RUnlock()
	return a.actuatorStates
}

// SetActuatorState manually sets actuator state for synchronization
func (a *ArduinoSerial) SetActuatorState(actuator string, value interface{}) {
	a.mutex.Lock()
	defer a.mutex.Unlock()

	switch actuator {
	case "white_light":
		if val, ok := value.(bool); ok {
			a.actuatorStates.WhiteLight = val
		}
	case "yellow_light":
		if val, ok := value.(bool); ok {
			a.actuatorStates.YellowLight = val
		}
	case "relay":
		if val, ok := value.(bool); ok {
			a.actuatorStates.Relay = val
		}
	case "fan":
		if val, ok := value.(bool); ok {
			a.actuatorStates.Fan = val
		}
	case "buzzer":
		if val, ok := value.(bool); ok {
			a.actuatorStates.Buzzer = val
		}
	case "door_angle":
		if val, ok := value.(float64); ok {
			a.actuatorStates.DoorAngle = int(val)
		}
	case "window_angle":
		if val, ok := value.(float64); ok {
			a.actuatorStates.WindowAngle = int(val)
		}
	case "fan_speed":
		if val, ok := value.(float64); ok {
			a.actuatorStates.FanSpeed = int(val)
		}
	}
}

// startDataListener listens for incoming data from Arduino
func (a *ArduinoSerial) startDataListener() {
	scanner := bufio.NewScanner(a.port)

	for scanner.Scan() {
		select {
		case <-a.stopChan:
			return
		default:
			line := strings.TrimSpace(scanner.Text())
			if line != "" {
				log.Printf("← %s", line)
				a.handleIncomingData(line)
			}
		}
	}

	if err := scanner.Err(); err != nil {
		log.Printf("Serial scanner error: %v", err)
	}
}

// startDataSaver periodically saves buffered data
func (a *ArduinoSerial) startDataSaver() {
	ticker := time.NewTicker(a.saveInterval)
	defer ticker.Stop()

	for {
		select {
		case <-a.stopChan:
			return
		case <-ticker.C:
			a.saveBufferedData()
		}
	}
}

// handleIncomingData processes incoming serial data
func (a *ArduinoSerial) handleIncomingData(line string) {
	// Handle ACK
	if line == "ACK" {
		a.mutex.Lock()
		a.isReady = true

		// Send next queued command
		if len(a.commandQueue) > 0 {
			nextCommand := a.commandQueue[0]
			a.commandQueue = a.commandQueue[1:]
			a.mutex.Unlock()

			// Send with slight delay
			time.AfterFunc(100*time.Millisecond, func() {
				a.SendCommand(nextCommand)
			})
		} else {
			a.mutex.Unlock()
		}
		return
	}

	// Parse sensor data
	if strings.Contains(line, "GAS:") {
		a.parseSensorData(line)
	} else {
		// Parse actuator responses
		a.parseActuatorResponse(line)
	}
}

// parseSensorData parses sensor data from Arduino
func (a *ArduinoSerial) parseSensorData(line string) {
	// Format: "GAS:123,LIGHT:456,SOIL:789,WATER:101,INFRAR:1,BTN1:0,BTN2:1"
	data := &models.SensorReading{
		Timestamp: time.Now(),
	}

	parts := strings.Split(line, ",")
	for _, part := range parts {
		keyValue := strings.Split(part, ":")
		if len(keyValue) != 2 {
			continue
		}

		key := strings.ToLower(strings.TrimSpace(keyValue[0]))
		valueStr := strings.TrimSpace(keyValue[1])
		value, err := strconv.Atoi(valueStr)
		if err != nil {
			continue
		}

		switch key {
		case "gas":
			data.Gas = value
		case "light":
			data.Light = value
		case "soil":
			data.Soil = value
		case "water":
			data.Water = value
		case "infrar":
			data.Infrar = value
		case "btn1":
			data.Btn1 = value
		case "btn2":
			data.Btn2 = value
		}
	}

	// Validate that we have the main sensor readings
	if data.Gas == 0 && data.Light == 0 && data.Soil == 0 && data.Water == 0 {
		return
	}

	a.mutex.Lock()
	a.currentData = data

	// Add to buffer for saving
	sensorData := models.SensorData{
		Light:     data.Light,
		Gas:       data.Gas,
		Soil:      data.Soil,
		Water:     data.Water,
		Infrared:  data.Infrar,
		Timestamp: data.Timestamp,
	}

	// Evaluate rules and get alerts
	alerts := a.ruleService.EvaluateRules(data)
	if len(alerts) > 0 {
		sensorData.Alerts = alerts

		// Execute triggered actions
		a.executeTriggeredActions(alerts)
	}

	a.dataBuffer = append(a.dataBuffer, sensorData)

	// Keep buffer size manageable
	if len(a.dataBuffer) > 100 {
		a.dataBuffer = a.dataBuffer[len(a.dataBuffer)-50:]
	}

	a.mutex.Unlock()
}

// parseActuatorResponse parses actuator state changes from Arduino
func (a *ArduinoSerial) parseActuatorResponse(line string) {
	a.mutex.Lock()
	defer a.mutex.Unlock()

	line = strings.ToLower(line)

	// Parse various actuator responses
	if strings.Contains(line, "light on") {
		if strings.Contains(line, "white") {
			a.actuatorStates.WhiteLight = true
		} else if strings.Contains(line, "yellow") {
			a.actuatorStates.YellowLight = true
		}
	} else if strings.Contains(line, "light off") {
		if strings.Contains(line, "white") {
			a.actuatorStates.WhiteLight = false
		} else if strings.Contains(line, "yellow") {
			a.actuatorStates.YellowLight = false
		}
	} else if strings.Contains(line, "fan on") {
		a.actuatorStates.Fan = true
	} else if strings.Contains(line, "fan off") {
		a.actuatorStates.Fan = false
	} else if strings.Contains(line, "relay on") {
		a.actuatorStates.Relay = true
	} else if strings.Contains(line, "relay off") {
		a.actuatorStates.Relay = false
	} else if strings.Contains(line, "buzzer on") {
		a.actuatorStates.Buzzer = true
	} else if strings.Contains(line, "buzzer off") {
		a.actuatorStates.Buzzer = false
	}

	// Parse servo positions
	if matches := regexp.MustCompile(`door.*?(\d+)`).FindStringSubmatch(line); len(matches) > 1 {
		if angle, err := strconv.Atoi(matches[1]); err == nil {
			a.actuatorStates.DoorAngle = angle
		}
	}
	if matches := regexp.MustCompile(`window.*?(\d+)`).FindStringSubmatch(line); len(matches) > 1 {
		if angle, err := strconv.Atoi(matches[1]); err == nil {
			a.actuatorStates.WindowAngle = angle
		}
	}
}

// executeTriggeredActions executes actions based on triggered rules
func (a *ArduinoSerial) executeTriggeredActions(alerts []string) {
	// Extract actions from alerts and execute them
	for _, alert := range alerts {
		// Simple action extraction - in production, you'd parse this more robustly
		if strings.Contains(alert, "buzzer_on") {
			a.SendCommand("BUZZER_ON")
		} else if strings.Contains(alert, "white_light_on") {
			a.SendCommand("WHITE_LIGHT_ON")
		} else if strings.Contains(alert, "window_close") {
			a.SendCommand("WINDOW_CLOSE")
		}
	}
}

// saveBufferedData saves buffered sensor data to database
func (a *ArduinoSerial) saveBufferedData() {
	a.mutex.Lock()
	if len(a.dataBuffer) == 0 {
		a.mutex.Unlock()
		return
	}

	// Copy buffer and clear it
	dataToSave := make([]models.SensorData, len(a.dataBuffer))
	copy(dataToSave, a.dataBuffer)
	a.dataBuffer = a.dataBuffer[:0]
	a.mutex.Unlock()

	// Save to database
	if err := a.sensorService.SaveBulkSensorData(dataToSave); err != nil {
		log.Printf("Error saving sensor data: %v", err)

		// Put data back in buffer if save failed
		a.mutex.Lock()
		a.dataBuffer = append(dataToSave, a.dataBuffer...)
		a.mutex.Unlock()
	}
}
